src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	return ring, nil
+}
+
+// GetIndex returns the index of the Member associated with key. A key belongs to the first Member whose hash is
+// greater than the key. If no Members are greater, then the Ring wraps around to the first Member.
+func (ring *Ring) GetIndex(key string) int {
+	keyHash := ring.hashFn([]byte(key))
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= keyHash
+	})
+
+	// wrap around to the first node in the Ring.
+	if index >= len(ring.members) {
+		index = 0
+	}
+
+	return index
+}
+
+// Get is a convenience wrapper around GetIndex that returns the Member that associates to a key.
+func (ring *Ring) Get(key string) Member {
+	return ring.members[ring.GetIndex(key)]
+}
+
+// Lock attempts to lock a key in etcd using consistent hashing. If key does not belong to the ring's node,
+// the call blocks until context is cancelled or expires.
+func (ring *Ring) Lock(ctx context.Context, key string) (context.Context, error) {
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+	for {
+		index := ring.GetIndex(key)
+		if ring.node.Id == ring.members[index].Id {
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, errors.EnsureStack(ctx.Err())
+		case <-ticker.C:
+		}
+	}
+
+	if info, alreadyExists := ring.node.Locks.Load(key); alreadyExists {
+		lockCtx := info.(lockInfo).Ctx
+		return lockCtx, nil
+	}
+
+	nodeLock := dlock.NewDLock(ring.Client, key)
+	lockCtx, err := nodeLock.Lock(ctx)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	ring.node.Locks.Store(key, lockInfo{
+		Lock: nodeLock,
+		Ctx:  lockCtx,
+	})
+
+	ring.Logger.WithFields(ring.lockLogFields(key)).Info("claimed lock")
+
+	return lockCtx, nil
+}
+
+// Unlock attempts to unlock a key in etcd if and only if a ring's node owns a lock reference to it.
+// Remote Member instances must give up their own locks.
+func (ring *Ring) Unlock(key string) error {
+	info, localLock := ring.node.Locks.Load(key)
+	if !localLock {
+		return nil // lock does not exist in this local process.
+	}
+
+	lockInfo := info.(lockInfo)
+	if err := lockInfo.Lock.Unlock(lockInfo.Ctx); err != nil {
+		return errors.EnsureStack(err)
+	}
+
+	ring.node.Locks.Delete(key)
+
+	ring.Logger.WithFields(ring.lockLogFields(key)).Info("released lock")
+
+	return nil
+}
+
+// start kicks off watch(), creates a node, and renews its lease.
+// todo: replace with collections stuff.
+func (ring *Ring) start(id string) error {
+	etcdCol := collection.NewEtcdCollection(ring.Client, path.Join(ring.Prefix, "nodes"),
+		nil, &types.BoolValue{}, nil, nil)
+
+	go func() {
+		defer ring.Cancel()
+		ring.watch(etcdCol)
+	}()
+
+	if err := ring.createNode(id, etcdCol); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createNode creates a lease, inserts itself as a key to etcd, keeps the lease alive in the background.
+// it is not idempotent, so if two rings attempt to add the same 'id', the second add will fail. This should
```

Comment:
UUID collisions aren't a realistic concern, so we can simplify this function a bit. I think the setup can happen outside of the callback since we can be confident there isn't a collision, so we don't need the ready channel. I think the setup should be in `New`, similar to the other comments, and the claim can be in a function that receives a context and is kicked off as a goroutine in the `New` function. I generally prefer setup to be simple enough to be captured in the `New` function and goroutines kicked off there so it is obvious what goroutines are running when you create the abstraction.